[PIO-61] Add S3 Model Data Repository

marevol opened the pull request.

dszeto commented on storage/s3/build.sbt:
```diff
@@ -0,0 +1,44 @@
+assemblyShadeRules in assembly := Seq(
+  ShadeRule.rename("org.apache.http.**" -> "shadeio.data.s3.http.@1").inAll,
+  ShadeRule.rename("com.fasterxml.**" -> "shadeio.data.s3.fasterxml.@1").inAll
+)
```
Hey @marevol can you explain a little bit regarding these shade rules, please?

marevol replied:
aws-java-sdk-s3 depends on them, but they conflict with different version files in spark-assembly jar.
To avoid this conflict, the package replacement is needed in storage/s3.

dszeto replied:
I think @mars is seeing some issues with https://github.com/apache/incubator-predictionio/blob/develop/storage/elasticsearch/build.sbt#L42.

marevol replied:
@mars could you provide info?
ES storage also conflicts with httpclient jar file of Spark.
I think an exception occurs if removing it, so it will be needed.

mars replied:
Hi @marevol,

We're building a [predictionio-incubating branch to add authentication to the Elasticsearch REST client](https://github.com/apache/incubator-predictionio/pull/372) with Scala 2.11 & Spark 2.1.

I believe it was during a runtime ES REST client call to `performRequest` suddenly an underlying method signature could not be located in `org.apache.http` package. @dszeto helped me to discover that the `shadeio.data` name was being incorrectly resolved. So, I tried [removing the shade rule](https://github.com/apache/incubator-predictionio/pull/372/files#diff-55cfeb297edd310e1efa8b6ac8bdbae6L39) and it started working. No more errors for that branch.

Maybe the difference is that we're using Spark 2.1?

marevol replied:
Thank you for the info.
My environment is Scala 2.10 and Spark 1.6.3.
I'll check it on Spark 2.1.

marevol replied:
Checking on Scala 2.11/Spark 2.1/ES 5.1&5.3, I could not reproduce it yet...
My test script file is [test_all.sh](https://github.com/jpioug/incubator-predictionio/blob/express-v1/test_all.sh) to build pio and run recommender template.
Is it reproduced in Docker tests? I'd like to know steps to reproduce it...

dszeto commented:
@marevol Would you like to add some integration test for this? https://github.com/jubos/fake-s3 or https://github.com/atlassian/localstack might be useful for mocking S3 locally within Travis.

marevol commented:
Added LocalStack for S3 testing.
LocalStack/moto seems to have a problem about ETag handling...
I added DISABLE_CHUNKED_ENCODING option to avoid it in S3 testing.

marevol commented:
I would like to merge S3 support this week.
For travis testing, it passed in https://travis-ci.org/jpioug/incubator-predictionio/builds/235569236 though apache-travis was failed...
Please let me know if you have any comments/concerns.

dszeto commented:
Sorry for the delay. I was on vacation.

This looks great to me, especially the use of `localstack`. I think the Travis failed downloading a Python library and erred out. It shouldn't matter IMO. I'll merge this now.
